### PR TITLE
Features/v6/tasks

### PR DESCRIPTION
**Pull request by Sauilitired**
Rewrite the task system.

Done so far:
- Move from ticks to a generic task time
- Add time converters to convert from real time to server ticks
- Move away from task IDs and instead use a task type that can be cancelled
- Add documentation to all task methods
- Rework the sync methods

### CODE REVIEW

File: Bukkit/src/main/java/com/plotsquared/bukkit/inject/BukkitModule.java [hunk 1]
```diff
@@ -60,8 +64,10 @@
 import com.plotsquared.core.util.SchematicHandler;
 import com.plotsquared.core.util.SetupUtils;
 import com.plotsquared.core.util.WorldUtil;
+import com.plotsquared.core.util.task.TaskTime;
 import com.sk89q.worldedit.bukkit.WorldEditPlugin;
 import com.sk89q.worldedit.extension.platform.Actor;
+import io.papermc.lib.PaperLib;
 import lombok.RequiredArgsConstructor;
```
Comment (dordsor21) [inline Bukkit/src/main/java/com/plotsquared/bukkit/inject/BukkitModule.java:71]: Why're there just a bunch of imports added, with nothing else changed in the file?

File: Bukkit/src/main/java/com/plotsquared/bukkit/listener/ChunkListener.java [hunk 2]
```diff
@@ -254,16 +255,16 @@ public void onEntitySpawn(CreatureSpawnEvent event) {
     private void cleanChunk(final Chunk chunk) {
         TaskManager.index.incrementAndGet();
         final Integer currentIndex = TaskManager.index.get();
-        Integer task = TaskManager.runTaskRepeat(() -> {
+        PlotSquaredTask task = TaskManager.runTaskRepeat(() -> {
             if (!chunk.isLoaded()) {
-                Bukkit.getScheduler().cancelTask(TaskManager.tasks.get(currentIndex));
+                TaskManager.tasks.get(currentIndex).cancel();
```
Comment (dordsor21) [inline Bukkit/src/main/java/com/plotsquared/bukkit/listener/ChunkListener.java:260]: Not pretty, but I guess it's a step towards a "standard"

File: Bukkit/src/main/java/com/plotsquared/bukkit/util/task/PaperTimeConverter.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import com.plotsquared.core.util.task.TaskTime;
+import org.bukkit.Bukkit;
+
+import javax.annotation.Nonnegative;
+
+/**
+ * Time converter that uses the server MSPT count to convert between
+ * different time units
+ */
```
Comment (dordsor21) [inline Bukkit/src/main/java/com/plotsquared/bukkit/util/task/PaperTimeConverter.java:36]: I like this

File: Core/src/main/java/com/plotsquared/core/util/task/AutoClaimFinishTask.java [hunk 4]
```diff
@@ -36,22 +36,24 @@
 import com.plotsquared.core.util.EventDispatcher;
 import lombok.RequiredArgsConstructor;
 
+import java.util.concurrent.Callable;
+
 import static com.plotsquared.core.util.MainUtil.sendMessage;
 
 @RequiredArgsConstructor
-public final class AutoClaimFinishTask extends RunnableVal<Object> {
+public final class AutoClaimFinishTask implements Callable<Boolean> {
```
Comment (dordsor21) [inline Core/src/main/java/com/plotsquared/core/util/task/AutoClaimFinishTask.java:44]: I see you started on changing to Callable a little

File: Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java [hunk 5]
```diff
@@ -25,144 +25,227 @@
...
 public abstract class TaskManager {
 
     public static final HashSet<String> TELEPORT_QUEUE = new HashSet<>();
-    public static final HashMap<Integer, Integer> tasks = new HashMap<>();
+    public static final HashMap<Integer, PlotSquaredTask> tasks = new HashMap<>();
     public static AtomicInteger index = new AtomicInteger(0);
 
-    @Getter @Setter private static TaskManager implementation;
-
-    public static int runTaskRepeat(Runnable runnable, int interval) {
```
Comment (dordsor21) [inline Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java:48]: Do we still have the general plan to implement our own platform-agnostic Async stuff, for whatever might never use it?

File: Bukkit/src/main/java/com/plotsquared/bukkit/inject/BukkitModule.java (hunk 1 above)
Comment (Sauilitired) [reply Bukkit/src/main/java/com/plotsquared/bukkit/inject/BukkitModule.java:71]: Good question =)))

File: Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java (hunk 5 above)
Comment (Sauilitired) [reply Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java:48]: Maybe:tm: We currently don't need it, if we do, we implement it then. There's nothing stopping us. I also think it's fine if the static methods use the platform one, so we use the platform task manager by default.

File: Core/src/main/java/com/plotsquared/core/plot/expiration/ExpireManager.java [hunk 6]
```diff
@@ -376,9 +377,9 @@ public boolean runTask(final RunnableVal3<Plot, Runnable, Boolean> expiredTask)
                             ExpireManager.this.running = 2;
                             runTask(expiredTask);
                         }
-                    }, 86400000);
+                    }, TaskTime.ticks(86400000));
```
Comment (SirYwell) [inline Core/src/main/java/com/plotsquared/core/plot/expiration/ExpireManager.java:380]: You suffixed all other values with `L`

File: Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java [hunk 7]
```diff
@@ -25,144 +25,227 @@
...
+ * operations happen at the same time. Implementations of
+ * the task manager might make other guarantees. All asynchronous
+ * operations will happen on another thread, no matter where
+ * they're scheduled from.
+ */
 public abstract class TaskManager {
 
     public static final HashSet<String> TELEPORT_QUEUE = new HashSet<>();
-    public static final HashMap<Integer, Integer> tasks = new HashMap<>();
+    public static final HashMap<Integer, PlotSquaredTask> tasks = new HashMap<>();
```
Comment (SirYwell) [inline Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java:55]: Can we make this one private?

File: Bukkit/src/main/java/com/plotsquared/bukkit/util/task/PaperTimeConverter.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+public final class PaperTimeConverter implements TaskTime.TimeConverter {
+
+    @Override public long msToTicks(@Nonnegative final long ms) {
+        return (long) (ms / Bukkit.getAverageTickTime());
```
Comment (SirYwell) [inline Bukkit/src/main/java/com/plotsquared/bukkit/util/task/PaperTimeConverter.java:40]: What happens if this returns `0`?

File: Bukkit/src/main/java/com/plotsquared/bukkit/util/task/SpigotTimeConverter.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Naive time converter that assumes that all ticks are 50 milliseconds
+ */
+public final class SpigotTimeConverter implements TaskTime.TimeConverter {
+
+    private static final long MS_PER_TICKS = 50L;
+
+    @Override public long msToTicks(@Nonnegative final long ms) {
+        return ms / MS_PER_TICKS;
```
Comment (SirYwell) [inline Bukkit/src/main/java/com/plotsquared/bukkit/util/task/SpigotTimeConverter.java:40]: Same as for PaperTimeConverter

File: Core/src/main/java/com/plotsquared/core/command/Trim.java [hunk 10]
```diff
@@ -189,11 +191,12 @@ public static boolean getTrimRegions(String world,
...
                             }
                             final LocalBlockQueue queue = blockQueue.getNewQueue(world, false);
-                            TaskManager.objectTask(chunks, new RunnableVal<BlockVector2>() {
+                            TaskManager.getPlatformImplementation().objectTask(chunks, new RunnableVal<BlockVector2>() {
                                 @Override public void run(BlockVector2 value) {
                                     queue.regenChunk(value.getX(), value.getZ());
                                 }
-                            }, this);
+                            }).thenAccept(ignore -> TaskManager.getPlatformImplementation()
+                                .taskLater(this, TaskTime.ticks(1)));
```
Comment (SirYwell) [inline Core/src/main/java/com/plotsquared/core/command/Trim.java:199]: Missing `L` here too

File: Core/src/main/java/com/plotsquared/core/util/RegionManager.java [hunk 11]
```diff
@@ -94,12 +95,13 @@ public void largeRegionTask(final String world, final CuboidRegion region,
...
             }
-            TaskManager.objectTask(chunks, new RunnableVal<BlockVector2>() {
+            TaskManager.getPlatformImplementation().objectTask(chunks, new RunnableVal<BlockVector2>() {
                 @Override public void run(BlockVector2 value) {
                     chunkManager.loadChunk(world, value, false)
                         .thenRun(() -> task.run(value));
                 }
-            }, whenDone);
+            }).thenAccept(ignore ->
+                TaskManager.getPlatformImplementation().taskLater(whenDone, TaskTime.ticks(1)));
```
Comment (SirYwell) [inline Core/src/main/java/com/plotsquared/core/util/RegionManager.java:104]: Very important L missing here

File: Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java (hunk 7 above)
Comment (Sauilitired) [reply Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java:55]: then we need a getter for it ^-^

Comment (SirYwell) [reply Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java:55]: Or we don't expose it directly at all

File: Bukkit/src/main/java/com/plotsquared/bukkit/util/task/PaperTimeConverter.java (hunk 8 above)
Comment (Sauilitired) [reply Bukkit/src/main/java/com/plotsquared/bukkit/util/task/PaperTimeConverter.java:40]: the average tick time cannot be 0... that's impossible lol

File: Bukkit/src/main/java/com/plotsquared/bukkit/util/task/SpigotTimeConverter.java (hunk 9 above)
Comment (Sauilitired) [reply Bukkit/src/main/java/com/plotsquared/bukkit/util/task/SpigotTimeConverter.java:40]: 0/<non-zero> is fine, and MS_PER_TICKS is a constant.

File: Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java [hunk 12]
```diff
@@ -25,144 +25,269 @@
...
+    }
+
+    /**
+     * Remove a task from the task map and return the stored value
+     *
+     * @param id Task ID
+     * @return Task if stored, or {@code null}
+     */
+    @Nullable public static PlotSquaredTask removeTask(final int id) {
+        return tasks.get(id);
```
Comment (SirYwell) [inline Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java:99]: This only gets but doesn't remove the task

Comment (Sauilitired) [reply Core/src/main/java/com/plotsquared/core/util/task/TaskManager.java:99]: whoops lol

**Pull request merged**